File: firmware/cyd_desktop_board/modules/_manifest.py
"""The manifest.json schema for /sd/apps/<app-id>/.

One definition, used by the loader and the launcher. The same rules are
implemented for the PC side in bridge/manifest.mjs; both are checked against
firmware/cyd_desktop_board/tests/manifest-cases.json, and the schema is
documented in docs/cyd-api.md.

    {
      "id":      "weather",   optional; must equal the directory name
      "title":   "WEATHER",   required; 1-12 of A-Z a-z 0-9 space _ . -
      "entry":   "main.py",   optional; the .py file to run
      "api":     2,           optional; the cyd API level the app needs
      "version": "0.0.1"      optional; informational, up to 16 characters
    }

An app directory without manifest.json is valid: its title is the app id and
its entry is main.py. Errors are raised as ValueError("<field>: <reason>") so
callers can report which field was wrong.
"""

import json

# Bump when cyd gains functions an app may depend on. An app whose manifest asks
# for a higher level is refused instead of failing halfway through with an
# AttributeError.
API_LEVEL = 4

MAX_BYTES = 2048
KEYS = ("id", "title", "entry", "api", "version")


def safe_component(value, maximum=24):
    if not isinstance(value, str) or not value or len(value) > maximum or ".." in value:
        return False
    for character in value:
        if not (("a" <= character <= "z") or ("A" <= character <= "Z")
                or ("0" <= character <= "9") or character in "_.-"):
            return False
    return True


def safe_title(value):
    if not isinstance(value, str) or not value or len(value) > 12:
        return False
    for character in value:
        if not (("a" <= character <= "z") or ("A" <= character <= "Z")
                or ("0" <= character <= "9") or character in " _.-"):
            return False
    return True


def validate(app_id, data):
    """Return (title, entry) for an app, or raise ValueError("<field>: <reason>")."""
    if data is None:
        return app_id[:12], "main.py"
    if not isinstance(data, dict):
        raise ValueError("shape: manifest must be a JSON object")
    for key in data:
        if key not in KEYS:
            # Most often a typo, such as "tittle", that would otherwise be ignored.
            raise ValueError("key: unknown key " + str(key)[:16])
    if "id" in data and data["id"] != app_id:
        raise ValueError("id: does not match the directory name")
    title = data.get("title")
    if not safe_title(title):
        raise ValueError("title: need 1-12 of A-Z 0-9 space _.-")
    entry = data.get("entry", "main.py")
    if not safe_component(entry) or not entry.endswith(".py"):
        raise ValueError("entry: must be a .py file name")
    api = data.get("api", 1)
    if isinstance(api, bool) or not isinstance(api, int) or api < 1:
        raise ValueError("api: must be a positive integer")
    if api > API_LEVEL:
        raise ValueError("api: needs %d, firmware has %d" % (api, API_LEVEL))
    version = data.get("version")
    if version is not None and not (isinstance(version, str) and 0 < len(version) <= 16):
        raise ValueError("version: must be 1-16 characters")
    return title, entry


def load(app_id, root="/sd/apps"):
    """Read and validate an installed app's manifest."""
    path = root + "/" + app_id + "/manifest.json"
    try:
        with open(path, "r") as source:
            raw = source.read(MAX_BYTES + 1)
    except OSError:
        return validate(app_id, None)
    if len(raw) > MAX_BYTES:
        raise ValueError("shape: larger than %d bytes" % MAX_BYTES)
    try:
        data = json.loads(raw)
    except ValueError:
        raise ValueError("json: manifest.json is not valid JSON")
    return validate(app_id, data)

File: bridge/manifest.mjs
// PC-side check of an app's manifest.json, so a bad manifest is refused before
// anything is uploaded rather than making the app vanish from Scripts on the
// device.
//
// The schema lives in firmware/cyd_desktop_board/modules/_manifest.py. This is
// the same rules in JavaScript; both implementations are held to
// firmware/cyd_desktop_board/tests/manifest-cases.json. The API level is read
// out of _manifest.py so there is only one number to bump.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

const MANIFEST_MODULE = fileURLToPath(
  new URL('../firmware/cyd_desktop_board/modules/_manifest.py', import.meta.url));

export const KEYS = ['id', 'title', 'entry', 'api', 'version'];
export const MAX_MANIFEST_BYTES = 2048;

export const API_LEVEL = (() => {
  const match = /^API_LEVEL\s*=\s*(\d+)\s*$/m.exec(readFileSync(MANIFEST_MODULE, 'utf8'));
  if (!match) throw new Error(`API_LEVEL not found in ${MANIFEST_MODULE}`);
  return Number(match[1]);
})();

const COMPONENT = /^[A-Za-z0-9_.-]{1,24}$/;
const TITLE = /^[A-Za-z0-9 _.-]{1,12}$/;

export function isSafeComponent(value) {
  return typeof value === 'string' && COMPONENT.test(value) && !value.includes('..');
}

export function isSafeTitle(value) {
  return typeof value === 'string' && TITLE.test(value);
}

// Errors carry the offending field first, as "<field>: <reason>", matching the
// device so logs from either side read the same.
export class ManifestError extends Error {
  constructor(field, reason) {
    super(`${field}: ${reason}`);
    this.field = field;
  }
}

export function validateManifest(appId, data) {
  if (data === null || data === undefined) return { title: appId.slice(0, 12), entry: 'main.py' };
  if (typeof data !== 'object' || Array.isArray(data)) {
    throw new ManifestError('shape', 'manifest must be a JSON object');
  }
  for (const key of Object.keys(data)) {
    if (!KEYS.includes(key)) throw new ManifestError('key', `unknown key ${key.slice(0, 16)}`);
  }
  if ('id' in data && data.id !== appId) throw new ManifestError('id', 'does not match the directory name');
  if (!isSafeTitle(data.title)) throw new ManifestError('title', 'need 1-12 of A-Z 0-9 space _.-');
  const entry = data.entry ?? 'main.py';
  if (!isSafeComponent(entry) || !entry.endsWith('.py')) throw new ManifestError('entry', 'must be a .py file name');
  const api = data.api ?? 1;
  if (typeof api !== 'number' || !Number.isInteger(api) || api < 1) {
    throw new ManifestError('api', 'must be a positive integer');
  }
  if (api > API_LEVEL) throw new ManifestError('api', `needs ${api}, firmware has ${API_LEVEL}`);
  if (data.version !== undefined && data.version !== null
      && !(typeof data.version === 'string' && data.version.length > 0 && data.version.length <= 16)) {
    throw new ManifestError('version', 'must be 1-16 characters');
  }
  return { title: data.title, entry };
}

export function parseManifest(appId, text) {
  if (Buffer.byteLength(text, 'utf8') > MAX_MANIFEST_BYTES) {
    throw new ManifestError('shape', `larger than ${MAX_MANIFEST_BYTES} bytes`);
  }
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    throw new ManifestError('json', 'manifest.json is not valid JSON');
  }
  return validateManifest(appId, data);
}
